### PR TITLE
feat: add support for estimator eval hooks

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
We previously ignored hooks that were passed into `tf.estimator.EvalSpec`.

[DET-3097](https://determinedai.atlassian.net/browse/DET-3097)

## Test Plan
Tested manually, and also added a unit test for both training and eval hooks.

### CODE REVIEW

File: harness/determined/estimator/_estimator_trial.py [hunk 1]
```diff
@@ -500,7 +500,9 @@ def _init_model(self) -> None:
         repeating_train_fn = self._check_and_repeat_train_input_fn(self.user_train_spec.input_fn)
 
         self.train_spec = tf.estimator.TrainSpec(input_fn=repeating_train_fn, hooks=all_hooks)
-        self.eval_spec = tf.estimator.EvalSpec(input_fn=self.val_spec.input_fn, steps=None)
+        self.eval_spec = tf.estimator.EvalSpec(
+            input_fn=self.val_spec.input_fn, hooks=self.val_spec.hooks, steps=None
+        )
```
Comment (yoavz) [inline harness/determined/estimator/_estimator_trial.py:505]: that was easy!

File: harness/tests/experiment/fixtures/estimator_xor_model.py [hunk 2]
```diff
@@ -150,3 +151,45 @@ def build_validation_spec(self) -> tf.estimator.EvalSpec:
...
+        self._file_path = file_path
+        self._idx = 0
+
+    def before_run(self, run_context: Any) -> None:
+        self._idx += 1
+
+    def after_run(self, run_context: Any, run_values: Any) -> None:
+        with open(self._file_path, "w") as fp:
+            fp.write(f"{self._idx}")
+            fp.flush()
```
Comment (yoavz) [inline harness/tests/experiment/fixtures/estimator_xor_model.py:166]: non-blocking: i think the `flush()` is redundant since the file closes once the `with` block is exited.

File: harness/tests/experiment/tensorflow/test_estimator_trial.py [hunk 3]
```diff
@@ -197,6 +198,42 @@ def make_trial_controller_fn(
...
+            controller.run()
+
+            with open(hparams["training_log_path"], "r") as fp:
+                assert int(fp.readline()) == batches_per_step * steps
+
+            with open(hparams["val_log_path"], "r") as fp:
+                # Because we do not specify number of steps in validation, the first
+                # time we run validation, after going through all the examples, eval
+                # will call `before_run()` then hit an I/O exception. Thus we need
+                # to +1 to the expected number of calls.
```
Comment (yoavz) [inline harness/tests/experiment/tensorflow/test_estimator_trial.py:234]: blocking: i don't really understand this comment. if we hit an I/O exception on `before_run()` why doesn't this test fail? might want to add some more to this comment to help future maintainability, or just reframe the test in a way that it doesn't have this edge case.

File: harness/tests/experiment/fixtures/estimator_xor_model.py (hunk 2 above)
Comment (aaron276h) [reply harness/tests/experiment/fixtures/estimator_xor_model.py]: removed

File: harness/tests/experiment/tensorflow/test_estimator_trial.py (hunk 3 above)
Comment (aaron276h) [reply harness/tests/experiment/tensorflow/test_estimator_trial.py]: The I/O exception is used by Estimators to determined the end of the dataset on the first epoch if `max_steps` isn't specified (Estimators catches and handles it). For callbacks this means that on the last batch of the first epoch `before_run()` will be called one more than the number of data items present in the dataset, so we have to do + 1 on the expected result.

I simplified the test by moving the `self_idx += 1` to `after_run()` which is not called an extra time. We lose a bit of coverage on `before_run()` but should be ok.
